lms/services/canvas_api/factory.py

```diff
@@ -3,16 +3,37 @@
 from lms.services.canvas_api._basic import BasicClient
 from lms.services.canvas_api._pages import CanvasPagesClient
 from lms.services.canvas_api.client import CanvasAPIClient
+from lms.services.file import file_service_factory
+from lms.services.oauth2_token import oauth2_token_service_factory
 
 
-def canvas_api_client_factory(_context, request):
+def canvas_api_client_factory(
+    _context, request, application_instance=None, user_id=None
+):
     """
     Get a CanvasAPIClient from a pyramid request.
 
     :param request: Pyramid request object
     :return: An instance of CanvasAPIClient
     """
-    application_instance = request.lti_user.application_instance
+    if application_instance and user_id:
+        oauth2_token_service = oauth2_token_service_factory(
```

Comment:
Note that oauth2_token_service_factory already uses this pattern, takes AI and user_id.